content/blog/2019-11-13-trace-gen-llvm/index.md

```diff
@@ -0,0 +1,159 @@
+  %11 = load i32, i32* %3, align 4
+  ret i32 %11
+}
+```
+
+Phi Nodes should also be removed because there is no control flow to merge in the trace after block merging has occurred. When a Phi Node is removed, the single operand that exists should be forwarded to the uses of that Phi Node.
+
+We do not remove the basic blocks during the previously described traversal. It's possible a block that was not taken by a conditional branch may be jumped to later, so we can't delete it. Once block merging and inlining is complete, we delete all dead basic blocks by checking if any predecessors exist.
+
+The traces generated take the place of the given function. They do not contain guards that check if the path speculation was correct. Thus they must be run in isolation as a single loop iteration.
+
+## Evaluation
+
+We evaluate on a subset of benchmarks from [Embench](https://github.com/embench/embench-iot). These are single-threaded benchmarks targeted at embedded systems.
+
+We do not have an interpreter or language runtime so we could not inject trace code into the program, and importantly handle failed traces. Additionally, there are no guards to hook into this language runtime even if it did exist. For these reasons, we do not evaluate on entire programs, but rather small sections of the programs. Our methodology is a shown below.
+
+|         | Step | 
+| ------------- | ------------------- | 
+| 1 | Extract the body of a 'hot' loop into a function
+| 2 | Compile the function and run with the profiling pass
+| 3 | Compile the function with the trace generation pass
+| 4 | Run the generated trace
+
+In certain scenarios, we removed some loops (set the iterators to their initial values) and unrolled fixed size loops when possible.
+
+We evaluate on small code sections from four Embench benchmarks and one synthetic benchmark. The features of each code section are described below.
+
+| Benchmark | Conditional Branches | Function Calls 
+| ------------- |  ------------------- | ------------ |
+| minver | 2 | 1
+| aha-mont64 |  1 | 0
+| nbody | 0 | 1 (built-in, don't inline)
+| ud | 1 | 0
+| synthetic |2 (nested) | 0
+
+Our performance metric is the number of machine instructions generated by the optimized program vs. the un-optimized program in the observed function. We use `objdump -dC <binary>` and count the number of  instructions in the function. We compiled both at `-O0`, but included a dead code elimination (`--dce`) pass after both to eliminate no longer useful instructions from the trace (condition check and initializing un-used variables).
+
+| Benchmark | Un-opt Machine Inst. | Opt Machine Inst. | Inst. Reduction (%)
+| ------------- |  ------------------- | ------------ | ------- |
+| minver | 64 | 33 | 48
+| aha-mont64 |  45 | 34 | 24 
+| nbody | 130 | 130 | 0
+| ud | 56 | 50 | 11
+| synthetic | 39 | 12 | 69
+
+We see a reduction in the number of instructions generated for all programs with branches and inline-able functions.
```

Comment:
That's cool overall, but it would also have been interesting to see some manner of performance measurement. You can imagine that the instruction count might actually go _up_ for a traced function (because of unrolling) versus a more compact implementation with branches, but the performance could still be better.

Executing the programs would also have told you whether the tracing was "correct," i.e., it produced the same result as the un-traced original program.